### PR TITLE
EZP-28204: Change view data format to timestamp for TimeFieldType

### PR DESCRIPTION
**Pull request by webhdx**
> JIRA: https://jira.ez.no/browse/EZP-28204

# Description
This PR changes `eztime` form type data format to a timestamp which is easier to be handled by a JS datepicker components.

### CODE REVIEW

File: lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
+<?php
+
+/**
+ * This file is part of the eZ RepositoryForms package.
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ * @version //autogentag//
```
Comment (alongosz) [inline lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php]: Nitpick: please skip

File: lib/FieldType/DataTransformer/TimeValueTransformer.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+/**
+ * This file is part of the eZ RepositoryForms package.
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ * @version //autogentag//
```
Comment (alongosz) [inline lib/FieldType/DataTransformer/TimeValueTransformer.php]: Nitpick: skip

File: lib/Form/Type/FieldType/TimeFieldType.php [hunk 3]
```diff
@@ -44,21 +45,21 @@ public function getParent()
 
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
-        $builder->addModelTransformer(new FieldValueTransformer($this->fieldTypeService->getFieldType('eztime')));
+        $builder
+            ->addViewTransformer(new FormattedTimeToTimestampTransformer($options['with_seconds']))
```
Comment (alongosz) [inline lib/Form/Type/FieldType/TimeFieldType.php]: Are you sure `with_seconds` key will be always present?

File: lib/Form/Type/FieldType/TimeFieldType.php [hunk 4]
```diff
@@ -44,21 +45,21 @@ public function getParent()
...
+            ->addModelTransformer(new TimeValueTransformer());
     }
 
     public function buildView(FormView $view, FormInterface $form, array $options)
     {
-        if ($options['with_seconds']) {
-            $view->vars['attr']['step'] = 1;
-        }
+        $view->vars['attr']['data-seconds'] = (int) $options['with_seconds'];
```
Comment (alongosz) [inline lib/Form/Type/FieldType/TimeFieldType.php]: Ok, so is this option a `bool` (inferring from `FormattedTimeToTimestampTransformer` constructor call) or an `int`?
The "thing" on the other side, using view variables should be able to handle bool.
But I don't know if it's possible for forms, if not then this is fine.

File: lib/Form/Type/FieldType/TimeFieldType.php [hunk 5]
```diff
@@ -44,21 +45,21 @@ public function getParent()
 
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
-        $builder->addModelTransformer(new FieldValueTransformer($this->fieldTypeService->getFieldType('eztime')));
+        $builder
+            ->addViewTransformer(new FormattedTimeToTimestampTransformer($options['with_seconds']))
+            ->addModelTransformer(new TimeValueTransformer());
```
Comment (alongosz) [inline lib/Form/Type/FieldType/TimeFieldType.php]: Noob question: what's the difference between Model and View transformers in this case? Based on [Symfony Doc](http://symfony.com/doc/current/form/data_transformers.html#about-model-and-view-transformers) I'd expect both transformers to be Model transformers, but I don't know this that well (or rather at all :stuck_out_tongue_winking_eye:)

File: lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        if (empty($value)) {
+            return null;
+        }
+
+        if (!is_string($value)) {
+            throw new TransformationFailedException('Expected a string.');
+        }
+
+        $dateTime = DateTime::createFromFormat($this->withSeconds ? 'H:i:s' : 'H:i', $value);
```
Comment (mikadamczyk) [inline lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php]: Is it possible that the function `createFromFormat()` will return bool in this case?

File: lib/Form/Type/FieldType/TimeFieldType.php (hunk 5 above)
Comment (webhdx) [reply lib/Form/Type/FieldType/TimeFieldType.php]: There are 3 types of data in Symfony forms:
* Model data - used by your model
* Normalized - used by the form type itself
* View data - used as a value in HTML

Model Transformer is doing the conversion between Model<->Norm data.
View Transformer is doing the conversion between Norm<->View data.

File: lib/Form/Type/FieldType/TimeFieldType.php (hunk 4 above)
Comment (webhdx) [reply lib/Form/Type/FieldType/TimeFieldType.php]: It a `bool` value but we need to present is as either `0` or `1` as `data-*` HTML attributes do not support boolean values.

File: lib/Form/Type/FieldType/TimeFieldType.php (hunk 3 above)
Comment (webhdx) [reply lib/Form/Type/FieldType/TimeFieldType.php]: Yes. It has default value always set.

File: lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php (hunk 6 above)
Comment (webhdx) [reply lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php]: I don't think so? Why would it return `true`? It should return a proper object or `false` (gotta love PHP for that inconsistency) if the `$value` cannot be parsed using provided datetime format.

Comment (mikadamczyk) [reply lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php]: If it will return `false` for some reason, then next line will fail

Comment (MarioBlazek) [reply lib/FieldType/DataTransformer/FormattedTimeToTimestampTransformer.php]: This line will fail if you provide timestamp in `$value`.

**Comment by webhdx** [comment]
@Nattfarinn @alongosz @mikadamczyk Please note I changed my implementation. I over engineered it previously. There was no need to inherit `TimeType` and force it to use `DateTime` as a model data format (it was kinda hackish). Simple `IntegerType` will do the job. Sorry for the inconvenience.

**Pull request merged**